src/Task/TaskScheduler.php

```diff
@@ -0,0 +1,226 @@
+                    return;
+                }
+            }
+        }
+    }
+
+    private function removeThread(ParallelTaskInterface $thread): void
+    {
+        $this->output->writeln(sprintf(self::LOG_END, $thread->getToolName()), OutputInterface::VERBOSITY_DEBUG);
+        $report = $this->threads->offsetGet($thread);
+        $this->threads->detach($thread);
+        $this->runningThreads--;
+        $this->success = $this->success && $report->getStatus() === ToolReportInterface::STATUS_PASSED;
+        if ($this->fastFinish && !$this->success) {
+            $this->stop = true;
+        }
+    }
+
+    private function fillUp(): void
+    {
+        if ($this->stop || $this->runningThreads === $this->parallelThreads) {
+            // No slot available.
+            return;
+        }
+
+        if (!$this->tasks->valid()) {
+            // If we have no pending tasks left.
+            return;
+        }
+
+        while ($this->runningThreads < $this->parallelThreads && $this->tasks->valid()) {
+            $next = $this->tasks->current();
+            // If the pending task is not parallelizable, return if we still have some running.
+            if (! ($next instanceof ParallelTaskInterface) && $this->runningThreads > 0) {
+                return;
+            }
+            $this->enqueueTask($next);
+            $this->tasks->next();
+            if ($this->stop) {
+                break;
+            }
+        }
+    }
+
+    private function enqueueTask(ReportWritingTaskInterface $task): void
+    {
+        // Start it up.
+        $name = $task->getToolName();
+        $this->output->writeln(sprintf(self::LOG_START, $name), OutputInterface::VERBOSITY_DEBUG);
+        $report = $this->report->addToolReport($name);
+        try {
+            $task->runWithReport($report);
+        } catch (PluginApiRuntimeException $throwable) {
+            $this->renderException($this->output, $throwable);
+            if ($this->fastFinish) {
+                $this->stop = true;
+            }
+            return;
+        }
+
+        // We could not run this task parallel, do not attach.
+        if (!$task instanceof ParallelTaskInterface) {
+            $this->success = $this->success && $report->getStatus() === ToolReportInterface::STATUS_PASSED;
+            if ($this->fastFinish && !$this->success) {
+                $this->stop = true;
+            }
+
+            $this->output->writeln(sprintf(self::LOG_END, $name), OutputInterface::VERBOSITY_DEBUG);
+            return;
+        }
+        $this->threads->attach($task, $report);
+        $this->runningThreads++;
+    }
+
+    /**
+     * @return ReportWritingTaskInterface[]|Generator
+     * @psalm-return Generator<int, ReportWritingTaskInterface>
+     */
+    private function nextTaskGenerator(TasklistInterface $tasks): Generator
+    {
+        foreach ($tasks->getIterator() as $task) {
+            assert($task instanceof ReportWritingTaskInterface);
```

Comment:
Could you get rid of the assertion when using `foreach ($tasks as $task)`?

Reply:
I guess so, but then there is no (activatable) type check here anymore at all which might cause hard to detect problems.